# Исследование сервиса для чтения книг 

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Выведем-первые-строки-каждой-таблицы-и-изучим-значение-столбцов" data-toc-modified-id="Выведем-первые-строки-каждой-таблицы-и-изучим-значение-столбцов-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Выведем первые строки каждой таблицы и изучим значение столбцов</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-1.1.5"><span class="toc-item-num">1.1.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li></ul></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Кол-во-книг-вышедших-после--1-января-2000-года" data-toc-modified-id="Кол-во-книг-вышедших-после--1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Кол-во книг вышедших после  1 января 2000 года</a></span></li><li><span><a href="#Кол-во-обзоров-и-средняя-оценка-книги" data-toc-modified-id="Кол-во-обзоров-и-средняя-оценка-книги-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Кол-во обзоров и средняя оценка книги</a></span></li><li><span><a href="#Издательство-с-наибольшим-числом-выпущенных-книг-толще-50-страниц" data-toc-modified-id="Издательство-с-наибольшим-числом-выпущенных-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Издательство с наибольшим числом выпущенных книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

**Описание проекта**

Наша компания купила крупный сервис для чтения книг по подписке. В рамках данного проекта мы должны проанализировать предоставленную  базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Предоставленные файлы**

Схема данных
![База данных](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)


- `books` - таблица содержит данные о книгах
- `authors` - таблица содержит данные об авторах
- `publishers` - таблица содержит данные об издательствах
- `ratings` - таблица содержит данные о пользовательских оценках книг
- `reviews` - таблица содержит данные о пользовательских обзорах

**Цель исследования**

- Написать рабочие, оптимальные запросы к БД
- Проанализировать базу данных
- Сформулировать выводы

**Основные шаги**

1. Обзор данных
2. Исследовательский анализ данных
3. Общий вывод  и формулировка ценностного предложения для нового продукта

## Обзор данных

In [1]:
# импортируем библиотеки

import pandas as pd
import sqlalchemy as sa

Подключение к базе данных `data-analyst-final-project-db`

In [2]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}

connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:    
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:    
        return pd.read_sql(sql=sa.text(query), con = con)

### Выведем первые строки каждой таблицы и изучим значение столбцов

#### Таблица books

In [3]:
query = '''
SELECT *
FROM books
LIMIT 5'''
get_sql_data(query)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query = '''
SELECT COUNT(*)
FROM books
'''
get_sql_data(query)

count
0   1000

В таблице 6-ть столбцов и 1000 строк:
- **book_id** — идентификатор книги
- **author_id** — идентификатор автора 
- **title** — название книги
- **num_pages** — количество страниц
- **publication_date** — дата публикации книги
- **publisher_id** — идентификатор издателя

#### Таблица authors

In [5]:
query = '''
SELECT *
FROM authors
LIMIT 5'''

get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query = '''
SELECT COUNT(*)
FROM authors
'''
get_sql_data(query)

count
0    636

В таблице 2-а столбца и 636 строк:
- **author_id** — идентификатор автора
- **author** — имя автора

#### Таблица publishers

In [7]:
query = '''
SELECT *
FROM publishers
LIMIT 5'''

get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
query = '''
SELECT COUNT(*)
FROM publishers
'''
get_sql_data(query)

count
0    340

В таблице 2-а столбца и 340 строк:
- **publisher_id** — идентификатор издательства
- **publisher** — название издательства

#### Таблица ratings

In [9]:
query = '''
SELECT *
FROM ratings
LIMIT 5'''

get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query = '''
SELECT COUNT(*)
FROM ratings
'''
get_sql_data(query)

count
0   6456

В таблице 4-е столбца и 6456 строк:
- **rating_id** — идентификатор оценки
- **book_id** — идентификатор книги
- **username** — имя пользователя, оставившего оценку
- **rating** — оценка книги

#### Таблица reviews

In [11]:
query = '''
SELECT *
FROM reviews
LIMIT 5'''

get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
query = '''
SELECT COUNT(*)
FROM reviews
'''
get_sql_data(query)

count
0   2793

В таблице 4-е столбца и 2793 строк:
- **review_id** — идентификатор обзора
- **book_id** — идентификатор книги
- **username** — имя автора обзора
- **text** — текст обзора

**Вывод** 

В представленных таблицах - информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Предварительно данных достаточно для анализа.


## Исследовательский анализ данных

В рамках дальнейшего анализа изучим и проверим полученные данные. Найдём особенности. Соберём наблюдения в один общий вывод.

### Кол-во книг вышедших после  1 января 2000 года 

В рамках  исследования посчитаем, сколько книг вышло после 1 января 2000 года;

In [13]:
query = '''SELECT 
COUNT (*) AS book_count
FROM books 
WHERE publication_date > ('2000-01-01')'''
get_sql_data(query)

book_count
0         819

Кол-во книг вышедших после 1 января 2000 года составило 819 ед.

### Кол-во обзоров и средняя оценка книги

Для каждой книги посчитаем количество обзоров и среднюю оценку

In [14]:
query = '''
WITH first_sub_query AS
  (SELECT re.book_id,
          COUNT(re.review_id) AS review_count
   FROM reviews AS re
   GROUP BY re.book_id),
     second_sub_query AS
  (SELECT ra.book_id,
          ROUND(AVG(ra.rating), 2)AS rating_avg
   FROM ratings AS ra
   GROUP BY ra.book_id)
SELECT b.book_id,
       b.title,
       first_sub_query.review_count,
       second_sub_query.rating_avg
FROM books AS b
LEFT JOIN first_sub_query ON b.book_id = first_sub_query.book_id
LEFT JOIN second_sub_query ON b.book_id = second_sub_query.book_id
'''
get_sql_data(query)

book_id                                              title  review_count  \
0        652          The Body in the Library (Miss Marple  #3)           2.0   
1        273                                          Galápagos           2.0   
2         51                           A Tree Grows in Brooklyn           5.0   
3        951  Undaunted Courage: The Pioneering First Missio...           2.0   
4        839                                        The Prophet           4.0   
..       ...                                                ...           ...   
995       64                                Alice in Wonderland           4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)           2.0   
997      148                                          Christine           3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)           2.0   
999      828                           The Plot Against America           2.0   

     rating_avg  
0          4.50  
1          4.50  
2          4.25  
3          4.00  
4          4.29  
..          ...  
995        4.23  
996        5.00  
997        3.43  
998        3.50  
999        3.00  

[1000 rows x 4 columns]

По каждой книге найдено количество обзоров и средняя оценка

### Издательство с наибольшим числом выпущенных книг толще 50 страниц

Определим издательство, которое выпустило наибольшее число книг толще
50 страниц — так мы исключим из анализа брошюры

In [15]:
query = '''
SELECT p.publisher
FROM
  (SELECT b.publisher_id,
          COUNT(b.book_id) AS count_book
   FROM books AS b
   WHERE b.num_pages> 50
   GROUP BY b.publisher_id) AS sub_query
INNER JOIN publishers AS p ON sub_query.publisher_id = p.publisher_id
ORDER BY sub_query.count_book DESC
LIMIT 1
 '''

get_sql_data(query)

publisher
0  Penguin Books

Издательство с наибольшим числом выпущенных книг толще 50 страниц - **Penguin Books**

### Автор с самой высокой средней оценкой книг

Найдём автора с самой высокой средней оценкой книг. Учтём только книги с 50 и более оценками

In [16]:
query = ''' 
WITH first_sub_query AS
  (SELECT r.book_id
   FROM ratings AS r
   GROUP BY r.book_id
   HAVING COUNT(r.rating)>=50),
     second_sub_query AS
  (SELECT b.author_id,
          AVG(r.rating) AS avg_rating
   FROM first_sub_query
   INNER JOIN books AS b ON first_sub_query.book_id = b.book_id
   INNER JOIN ratings AS r ON b.book_id = r.book_id
   GROUP BY b.author_id
   ORDER BY avg_rating DESC
   LIMIT 1)
SELECT a.author,
       second_sub_query.author_id,
       ROUND(second_sub_query.avg_rating, 2)
FROM second_sub_query
INNER JOIN authors AS a ON second_sub_query.author_id = a.author_id
'''
get_sql_data(query)

author  author_id  round
0  J.K. Rowling/Mary GrandPré        236   4.29

Автор с самой высокой средней оценкой книг **J.K. Rowling/Mary GrandPré**

### Среднее количество обзоров от пользователей

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [17]:
query = '''
WITH first_sub_query AS
  (SELECT ra.username AS username
   FROM ratings AS ra
   GROUP BY ra.username
   HAVING COUNT(ra.rating)>48),
     second_sub_query AS
  (SELECT first_sub_query.username,
          COUNT(re.review_id) AS count_review
   FROM first_sub_query
   LEFT JOIN reviews AS re ON first_sub_query.username = re.username
   GROUP BY first_sub_query.username)
SELECT ROUND(AVG(count_review), 0) AS avg_count_review
FROM second_sub_query

'''
get_sql_data(query)

avg_count_review
0              24.0

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок составило 24 ед.


## Общий вывод

Мы произвели анализ предоставленной базы данных. В представленных таблицах - информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Предварительно данных достаточно для анализа.

Найденные особенности по исследованным данным:

- Кол-во книг вышедших после 1 января 2000 года составило 819 ед.
- По всем книгам (1000 ед.) найдено количество обзоров и средняя оценка
- Издательство с наибольшим числом выпущенных книг толще 50 страниц - **Penguin Books**
- Автор с самой высокой средней оценкой книг -  **J.K. Rowling/Mary GrandPré**
- Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок составило 24 ед.

Пример ценностного предложения для нового продукта: `Откройте дверь в увлекательный мир современной литературы с нашим уникальным сервисом чтения книг по подписке. Погрузитесь в 800 свежих произведений, выпущенных после 2000 года, оцененных более чем 1000 пользователями. Наше партнерство с ведущим издательством Penguin Books обеспечит Вам доступ к самому качественному контенту.` 